File: capacitor/tessera-mrz/ios/Sources/TesseraMrzPlugin/MrzResultMapper.swift
import Capacitor
import Foundation
import Tessera

struct MappedDocument {
    let format: String
    let documentCode: String
    let issuingState: String
    let documentNumber: String
    let primaryIdentifier: String
    let secondaryIdentifiers: [String]
    let nationality: String
    let dateOfBirth: String?
    let sex: String?
    let dateOfExpiry: String?
    let optionalData: String?

    func toJSObject() -> JSObject {
        [
            "format": format,
            "documentCode": documentCode,
            "issuingState": issuingState,
            "documentNumber": documentNumber,
            "primaryIdentifier": primaryIdentifier,
            "secondaryIdentifiers": secondaryIdentifiers,
            "nationality": nationality,
            "dateOfBirth": jsValue(dateOfBirth),
            "sex": jsValue(sex),
            "dateOfExpiry": jsValue(dateOfExpiry),
            "optionalData": jsValue(optionalData),
        ]
    }
}

struct MappedValidation {
    let structurallyValid: Bool
    let documentNumberCheckDigit: Bool?
    let dateOfBirthCheckDigit: Bool?
    let dateOfExpiryCheckDigit: Bool?
    let optionalDataCheckDigit: Bool?
    let compositeCheckDigit: Bool?

    func toJSObject() -> JSObject {
        [
            "structurallyValid": structurallyValid,
            "documentNumberCheckDigit": jsValue(documentNumberCheckDigit),
            "dateOfBirthCheckDigit": jsValue(dateOfBirthCheckDigit),
            "dateOfExpiryCheckDigit": jsValue(dateOfExpiryCheckDigit),
            "optionalDataCheckDigit": jsValue(optionalDataCheckDigit),
            "compositeCheckDigit": jsValue(compositeCheckDigit),
        ]
    }
}

struct MappedValidationFailure {
    let code: String
    let field: String?

    func toJSObject() -> JSObject {
        ["code": code, "field": jsValue(field)]
    }
}

struct MappedScanResult {
    let status: String
    let document: MappedDocument
    let validation: MappedValidation
    let validationFailures: [MappedValidationFailure]
    let rawLines: [String]

    func toJSObject() -> JSObject {
        var object: JSObject = [
            "status": status,
            "document": document.toJSObject(),
            "validation": validation.toJSObject(),
            "raw": ["lines": rawLines] as JSObject,
        ]
        if status == "partial" {
            object["validationFailures"] = validationFailures.map { $0.toJSObject() }
        }
        return object
    }
}

enum MrzResultMapper {
    static func map(_ parse: ParseResult) -> MappedScanResult? {
        let document: MrzDocument
        let status: String
        if let success = parse as? ParseResult.Success {
            document = success.document
            status = "success"
        } else if let partial = parse as? ParseResult.PartialSuccess {
            document = partial.document
            status = "partial"
        } else {
            return nil
        }

        let failures = parse.metadata.validationFailures
        return MappedScanResult(
            status: status,
            document: mapDocument(document),
            validation: mapValidation(document, failures: failures),
            validationFailures: failures.map(mapFailure),
            rawLines: document.rawLines
        )
    }

    static func documentFamily(_ parse: ParseResult) -> String? {
        let document: MrzDocument
        if let success = parse as? ParseResult.Success {
            document = success.document
        } else if let partial = parse as? ParseResult.PartialSuccess {
            document = partial.document
        } else {
            return nil
        }

        switch document.format.name {
        case "TD3": return "passport"
        case "TD1", "TD2": return "id-card"
        case "MRV_A", "MRV_B": return "visa"
        default: return nil
        }
    }

    private static func mapDocument(_ document: MrzDocument) -> MappedDocument {
        let fields = document.commonFields
        return MappedDocument(
            format: document.format.name.replacingOccurrences(of: "MRV_", with: "MRV"),
            documentCode: fields.documentType as? String ?? "",
            issuingState: fields.issuingState as? String ?? "",
            documentNumber: withoutTrailingFiller(fields.documentNumber),
            primaryIdentifier: fields.primaryIdentifier,
            secondaryIdentifiers: fields.secondaryIdentifier.split(whereSeparator: \.isWhitespace).map(String.init),
            nationality: fields.nationality as? String ?? "",
            dateOfBirth: fields.dateOfBirth.computedDate?.description(),
            sex: character(fields.rawSex).flatMap { $0 == "<" ? nil : $0 },
            dateOfExpiry: fields.dateOfExpiry.computedDate?.description(),
            optionalData: optionalData(document).map(withoutTrailingFiller).flatMap { $0.isEmpty ? nil : $0 }
        )
    }

    private static func optionalData(_ document: MrzDocument) -> String? {
        if let td1 = document as? TD1 { return td1.optionalData1 + td1.optionalData2 }
        if let td2 = document as? TD2 { return td2.optionalData }
        if let td3 = document as? TD3 { return td3.personalNumber }
        if let mrvA = document as? MrvA { return mrvA.optionalData }
        if let mrvB = document as? MrvB { return mrvB.optionalData }
        return nil
    }

    private static func mapValidation(
        _ document: MrzDocument,
        failures: [MrzValidationError]
    ) -> MappedValidation {
        let mismatches = failures.compactMap { ($0 as? MrzCheckDigitMismatch)?.field }
        let checks = document.commonFields.checkDigits
        return MappedValidation(
            structurallyValid: true,
            documentNumberCheckDigit: !contains(mismatches, field: .documentNumber),
            dateOfBirthCheckDigit: !contains(mismatches, field: .dateOfBirth),
            dateOfExpiryCheckDigit: !contains(mismatches, field: .dateOfExpiry),
            optionalDataCheckDigit: checks.optionalData == nil ? nil : !contains(mismatches, field: .optionalData),
            compositeCheckDigit: checks.composite == nil ? nil : !contains(mismatches, field: .composite)
        )
    }

    private static func mapFailure(_ failure: MrzValidationError) -> MappedValidationFailure {
        let field: MrzField?
        if let mismatch = failure as? MrzCheckDigitMismatch {
            field = mismatch.field
        } else if let invalidDate = failure as? MrzDateNotInCalendar {
            field = invalidDate.field
        } else {
            field = nil
        }
        let className = NSStringFromClass(type(of: failure)).split(separator: ".").last.map(String.init)
        let code: String
        if let className, className.hasPrefix("Tessera") {
            code = String(className.dropFirst("Tessera".count))
        } else {
            code = className ?? "MrzValidationError"
        }
        return MappedValidationFailure(
            code: code,
            field: field.map(fieldName)
        )
    }

    private static func contains(_ fields: [MrzField], field: MrzField) -> Bool {
        fields.contains { $0 === field }
    }

    private static func fieldName(_ field: MrzField) -> String {
        if field === MrzField.documentType { return "document-type" }
        if field === MrzField.issuingState { return "issuing-state" }
        if field === MrzField.nameField { return "name" }
        if field === MrzField.documentNumber { return "document-number" }
        if field === MrzField.nationality { return "nationality" }
        if field === MrzField.dateOfBirth { return "date-of-birth" }
        if field === MrzField.dateOfExpiry { return "date-of-expiry" }
        if field === MrzField.optionalData { return "optional-data" }
        return "composite"
    }
}

private func withoutTrailingFiller(_ value: String) -> String {
    String(value.reversed().drop(while: { $0 == "<" }).reversed())
}

private func character(_ value: unichar) -> String? {
    guard let scalar = Unicode.Scalar(value) else { return nil }
    return String(Character(scalar))
}

private func jsValue(_ value: String?) -> JSValue {
    if let value { return value }
    return NSNull()
}

private func jsValue(_ value: Bool?) -> JSValue {
    if let value { return value }
    return NSNull()
}

File: capacitor/tessera-mrz/android/src/main/kotlin/io/lightine/tessera/capacitor/mrz/MrzResultMapper.kt
package io.lightine.tessera.capacitor.mrz

import io.lightine.tessera.mrz.model.MrvA
import io.lightine.tessera.mrz.model.MrvB
import io.lightine.tessera.mrz.model.MrzDocument
import io.lightine.tessera.mrz.model.TD1
import io.lightine.tessera.mrz.model.TD2
import io.lightine.tessera.mrz.model.TD3
import io.lightine.tessera.mrz.parsing.ParseResult
import io.lightine.tessera.types.errors.MrzCheckDigitMismatch
import io.lightine.tessera.types.errors.MrzDateNotInCalendar
import io.lightine.tessera.types.errors.MrzValidationError
import io.lightine.tessera.types.vocabulary.MrzField
import org.json.JSONArray
import org.json.JSONObject

internal data class MappedDocument(
    val format: String,
    val documentCode: String,
    val issuingState: String,
    val documentNumber: String,
    val primaryIdentifier: String,
    val secondaryIdentifiers: List<String>,
    val nationality: String,
    val dateOfBirth: String?,
    val sex: String?,
    val dateOfExpiry: String?,
    val optionalData: String?,
)

internal data class MappedValidation(
    val structurallyValid: Boolean,
    val documentNumberCheckDigit: Boolean?,
    val dateOfBirthCheckDigit: Boolean?,
    val dateOfExpiryCheckDigit: Boolean?,
    val optionalDataCheckDigit: Boolean?,
    val compositeCheckDigit: Boolean?,
)

internal data class MappedValidationFailure(
    val code: String,
    val field: String?,
)

internal data class MappedScanResult(
    val status: String,
    val document: MappedDocument,
    val validation: MappedValidation,
    val validationFailures: List<MappedValidationFailure>,
    val rawLines: List<String>,
) {
    fun toJson(): JSONObject =
        JSONObject().apply {
            put("status", status)
            put("document", document.toJson())
            put("validation", validation.toJson())
            if (status == "partial") {
                put("validationFailures", JSONArray(validationFailures.map { it.toJson() }))
            }
            put("raw", JSONObject().put("lines", JSONArray(rawLines)))
        }
}

internal object MrzResultMapper {
    fun map(parse: ParseResult): MappedScanResult? {
        val document =
            when (parse) {
                is ParseResult.Success -> parse.document
                is ParseResult.PartialSuccess -> parse.document
                is ParseResult.Failure -> return null
            }
        val failures = parse.metadata.validationFailures
        return MappedScanResult(
            status = if (parse is ParseResult.Success) "success" else "partial",
            document = mapDocument(document),
            validation = mapValidation(document, failures),
            validationFailures = failures.map(::mapFailure),
            rawLines = document.rawLines,
        )
    }

    fun documentFamily(parse: ParseResult): String? {
        val format =
            when (parse) {
                is ParseResult.Success -> parse.document.format.name
                is ParseResult.PartialSuccess -> parse.document.format.name
                is ParseResult.Failure -> return null
            }
        return when (format) {
            "TD3" -> "passport"
            "TD1", "TD2" -> "id-card"
            "MRV_A", "MRV_B" -> "visa"
            else -> null
        }
    }

    private fun mapDocument(document: MrzDocument): MappedDocument {
        val fields = document.commonFields
        return MappedDocument(
            format = document.format.name.replace("MRV_", "MRV"),
            documentCode = fields.documentType.rawCode,
            issuingState = fields.issuingState.rawCode,
            documentNumber = fields.documentNumber.trimEnd('<'),
            primaryIdentifier = fields.primaryIdentifier,
            secondaryIdentifiers = fields.secondaryIdentifier.split(Regex("\\s+")).filter(String::isNotBlank),
            nationality = fields.nationality.rawCode,
            dateOfBirth = fields.dateOfBirth.computedDate?.toString(),
            sex = fields.rawSex.takeUnless { it == '<' }?.toString(),
            dateOfExpiry = fields.dateOfExpiry.computedDate?.toString(),
            optionalData = optionalData(document)?.trimEnd('<')?.ifBlank { null },
        )
    }

    private fun optionalData(document: MrzDocument): String? =
        when (document) {
            is TD1 -> listOf(document.optionalData1, document.optionalData2).joinToString("")
            is TD2 -> document.optionalData
            is TD3 -> document.personalNumber
            is MrvA -> document.optionalData
            is MrvB -> document.optionalData
        }

    private fun mapValidation(
        document: MrzDocument,
        failures: List<MrzValidationError>,
    ): MappedValidation {
        val mismatches = failures.filterIsInstance<MrzCheckDigitMismatch>().map { it.field }.toSet()
        val checks = document.commonFields.checkDigits
        return MappedValidation(
            structurallyValid = true,
            documentNumberCheckDigit = MrzField.DOCUMENT_NUMBER !in mismatches,
            dateOfBirthCheckDigit = MrzField.DATE_OF_BIRTH !in mismatches,
            dateOfExpiryCheckDigit = MrzField.DATE_OF_EXPIRY !in mismatches,
            optionalDataCheckDigit = checks.optionalData?.let { MrzField.OPTIONAL_DATA !in mismatches },
            compositeCheckDigit = checks.composite?.let { MrzField.COMPOSITE !in mismatches },
        )
    }

    private fun mapFailure(failure: MrzValidationError): MappedValidationFailure =
        MappedValidationFailure(
            code = failure.javaClass.simpleName,
            field =
                when (failure) {
                    is MrzCheckDigitMismatch -> failure.field.toBridgeName()
                    is MrzDateNotInCalendar -> failure.field.toBridgeName()
                    else -> null
                },
        )
}

private fun MrzField.toBridgeName(): String =
    when (this) {
        MrzField.DOCUMENT_TYPE -> "document-type"
        MrzField.ISSUING_STATE -> "issuing-state"
        MrzField.NAME_FIELD -> "name"
        MrzField.DOCUMENT_NUMBER -> "document-number"
        MrzField.NATIONALITY -> "nationality"
        MrzField.DATE_OF_BIRTH -> "date-of-birth"
        MrzField.DATE_OF_EXPIRY -> "date-of-expiry"
        MrzField.OPTIONAL_DATA -> "optional-data"
        MrzField.COMPOSITE -> "composite"
    }

private fun MappedDocument.toJson(): JSONObject =
    JSONObject().apply {
        put("format", format)
        put("documentCode", documentCode)
        put("issuingState", issuingState)
        put("documentNumber", documentNumber)
        put("primaryIdentifier", primaryIdentifier)
        put("secondaryIdentifiers", JSONArray(secondaryIdentifiers))
        put("nationality", nationality)
        put("dateOfBirth", dateOfBirth ?: JSONObject.NULL)
        put("sex", sex ?: JSONObject.NULL)
        put("dateOfExpiry", dateOfExpiry ?: JSONObject.NULL)
        put("optionalData", optionalData ?: JSONObject.NULL)
    }

private fun MappedValidation.toJson(): JSONObject =
    JSONObject().apply {
        put("structurallyValid", structurallyValid)
        put("documentNumberCheckDigit", documentNumberCheckDigit ?: JSONObject.NULL)
        put("dateOfBirthCheckDigit", dateOfBirthCheckDigit ?: JSONObject.NULL)
        put("dateOfExpiryCheckDigit", dateOfExpiryCheckDigit ?: JSONObject.NULL)
        put("optionalDataCheckDigit", optionalDataCheckDigit ?: JSONObject.NULL)
        put("compositeCheckDigit", compositeCheckDigit ?: JSONObject.NULL)
    }

private fun MappedValidationFailure.toJson(): JSONObject =
    JSONObject().apply {
        put("code", code)
        put("field", field ?: JSONObject.NULL)
    }
